core/shared/src/main/scala/org/http4s/headers/From.scala

```diff
@@ -0,0 +1,21 @@
+package org.http4s
+package headers
+
+import cats.parse.Parser
+import org.http4s.internal.parsing.Rfc5322
+import org.typelevel.ci.CIStringSyntax
+
+object From extends HeaderCompanion[From]("From") {
+  override def parse(s: String): ParseResult[From] =
+    ParseResult.fromParser(parser, "Invalid From header")(s)
+
+  private[http4s] val parser: Parser[From] = Rfc5322.mailbox.map(From.apply)
+
+  implicit val headerInstance: Header[From, Header.Single] = Header.createRendered(
+    ci"From",
+    _.email,
+    parse,
+  )
+}
+
+final case class From(email: String)
```

Comment:
I don't think it should be allowed to create `From` from an arbitrary string, like this
```scala
val from = From("just an arbitrary string")
```
It is usually safer to make a separate constructor for that:
```
object From {
  def fromString(s: String): Option[From] = ???
}
```
Whereas `apply` has to be disabled.

There are a lot of examples on how to achieve it among other header models.

Reply:
I made the case class abstract to prevent creating new instances without going through the `parse` method.

Reply:
Personally, I am fine with `abstract sealed case class`. However, there were concerns with this approach as well, because in this case an `unapply` method will be auto-generated anyway which may be troublesome to maintain in the future. See discussion in https://github.com/http4s/http4s/pull/6544#issuecomment-1297510068.

So I'd like to double-check with @armanbilge if such a concern in applicable in this case.

Reply:
Thanks for the ping. Right, that's a good summary of the issue.

I don't have much experience with how HTTP spec/headers evolve. Is there any possibility it could get a new (optional) field in the future? If yes, then adding an `unapply` method would make it difficult to evolve our model compatibly.

Reply:
I adjusted the code similarly to https://github.com/http4s/http4s/pull/6544#issuecomment-1297647842